Change setView to respect user panning.

kentr opened the pull request:
PR for #149.

Tested on desktop and mobile (iOS).

I considered adding the `_userPanned` guard to the code that sets `_locateOnNextLocationFound` to `true`, but decided not to because I wasn't clear on the flow / logic.

kentr committed:
Change setView behavior to use _userPanned.

See https://github.com/domoritz/leaflet-locatecontrol/issues/149.

kentr committed:
Move _userPanned guard to top of drawMarker().

domoritz commented:
hmm, I tried it and it seems that setting  `setView=false` does not work. 

Moreover, can you also add an option `once`. That makes the map pan to the users location but does not pan the map when the location changes?

domoritz commented:
Also, the follow option is subsumed by your changes, right? 

Let's delete them then.

domoritz commented on src/L.Control.Locate.js:
```diff
@@ -396,6 +407,12 @@ You can find the project at: https://github.com/domoritz/leaflet-locatecontrol
+                this._map.on('dragstart', function (){
+                  // This listener is on map, not on locateControl.
+                  this.locateControl._userPanned = true;
+                });
```
this breaks. Please write

``` js
                this._map.on('dragstart', function () {
                  this._userPanned = true;
                }, this);
```

kentr commented:
> this breaks.

Ah, thanks for the correction.  I forgot that I had assigned the control object to `map.locateControl`.

> Also, the follow option is subsumed by your changes, right? 

You may be in a better position to answer that.  I see these pieces, which implies to me that `follow` still has an effect.  My interpretation is that `setView` does nothing if `follow == false`.

``` javascript
            if (this.options.follow && this._following) {
                this._locateOnNextLocationFound = true;
            }
```

``` javascript
            if (this.options.follow) {
                this._startFollowing(this._map);
            }
```

> Moreover, can you also add an option `once`. That makes the map pan to the users location but does not pan the map when the location changes?

So, `setView` can be either `false`, `always`, `once`, or `untilPan` (default)?

domoritz commented:
> You may be in a better position to answer that. I see these pieces, which implies to me that follow still has an effect. My interpretation is that setView does nothing if follow == false.

I admit that this is a bit confusing right now. I need to refactor this a bit. Following means that we set the view to the users location. I also had to look but it seems that the only way to not change the location of the marker when a new location is found is to set watch to false. 

Have a look at https://github.com/domoritz/leaflet-locatecontrol/blob/gh-pages/src/L.Control.Locate.js#L368

> So, setView can be either false, always, once, or untilPan (default)?

Yes.

kentr committed:
Many changes per PR discussion.

Per https://github.com/domoritz/leaflet-locatecontrol/pull/152.

Fix error with this.locateControl._userPanned = true. Move setView action code to separate doSetView() function. Add 'once' option. Confirm setView = false works as designed.

kentr commented:
@domoritz Please review.

To make the logic more clear, I moved the code which performs the `panTo()` into a handler function `doSetView()`.

Also:
- Added `'once'` option.
- Corrected the bug you found for `_userPanned`.
- Confirmed `setView: false` works as expected.

Regarding `follow`, I think that it's still required b/c it triggers `this._locateOnNextLocationFound` [in `_onLocationFound()`](https://github.com/kentr/leaflet-locatecontrol/blob/gh-pages/src/L.Control.Locate.js#L408).  I'm hesitant to remove it, but I put all of the `setView` guards together in `drawMarker()` so that it would be easier to separate them from `follow` when it's time to remove it.

domoritz commented:
Thank you! I have to carefully go through the changes.

kentr commented:
@domoritz Is there anything I can do to help this get merged?

domoritz commented:
I started working on this but have a deadline at the end of the week. I'll finish it up some time next week.
